30mins-become-contributor-of-tikv.md

```diff
@@ -0,0 +1,276 @@
+
+目前，由于 TiKV 的 builtin 函数尚未全部实现，对于无法下推的表达式，TiDB 只能自行解决。这无疑将成为提升 TiDB 速度的最大绊脚石。好消息是，TiKV 在实现 builtin 函数时，可以直接参考 TiDB 的对应函数逻辑（顺便可以帮 TiDB 找找 Bug），为我们减少了不少工作量。
+
+**Built-in 函数无疑是 TiDB 和 TiKV 成长道路上不可替代的一步，如此艰巨又庞大的任务，我们需要广大社区朋友们的支持与鼓励。亲爱的朋友们，想玩 rust 吗？想给 TiKV 提 PR 吗？想帮助 TiDB 跑得更快吗？动动您的小手指，拿 PR 来砸我们吧。您的 PR 一旦被采用，将会有小惊喜哦。**
+
+## 手把手教你实现 built-in 函数
+
+### Step 1: 准备下推函数
+
+在 TiKV 的 [https://github.com/pingcap/tikv/issues/3275](https://github.com/pingcap/tikv/issues/3275) issue 中，找到未实现的函数签名列表，选一个您想要实现的函数。
+
+### Step 2: 获取 TiDB 中可参考的逻辑实现
+
+在 TiDB 的 [expression](https://github.com/pingcap/tidb/tree/master/expression) 目录下查找相关 builtinXXXSig 对象，这里 XXX 为您要实现的函数签名，本例中以 [MultiplyIntUnsigned](https://github.com/pingcap/tikv/pull/3277) 为例，可以在 TiDB 中找到其对应的函数签名（builtinArithmeticMultiplyIntUnsignedSig）及 [实现](https://github.com/pingcap/tidb/blob/master/expression/builtin_arithmetic.go#L532)。
+
+### Step 3: 确定函数定义
+
+1. built-in 函数所在的文件名要求与 TiDB 的名称对应，如 TiDB 中，[expression](https://github.com/pingcap/tidb/tree/master/expression) 目录下的下推文件统一以 builtin_XXX 命名，对应到 TiKV 这边，就是 `builtin_XXX.rs`。若同名对应的文件不存在，则需要自行在同级目录下新建。对于本例，当前函数存放于 TiDB 的 [builtin_arithmetic.go](https://github.com/pingcap/tidb/blob/master/expression/builtin_arithmetic.go#L532) 文件里，对应到 TiKV 便是存放在 [builtin_arithmetic.rs](https://github.com/pingcap/tikv/blob/master/src/coprocessor/dag/expr/builtin_arithmetic.rs) 中。
+
+2. 函数名称：函数签名转为 rust 的函数名称规范，这里 MultiplyIntUnsigned 将会被定义为 `multiply_int_unsigned`。
+
+3. 函数返回值，可以参考 TiDB 中实现的 Eval 函数，对应关系如下：
+
+![表 1.png](https://upload-images.jianshu.io/upload_images/542677-1a493001f3c401ba.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)
+
+可以看到 TiDB 的 builtinArithmeticMultiplyIntUnsignedSig  对象实现了 evalInt 方法，故当前函数（`multiply_int_unsigned`）的返回类型应该为 `Result<Option<i64>>`。
+
+4. 函数的参数, 所有 builtin-in 的参数都与 Expression 的 eval 函数一致，即：
+
+	* 环境配置量 (ctx:&StatementContext)
+
+	* 该行数据每列具体值 ( row:&[Datum])
+
+综上，`multiply_int_unsigned` 的下推函数定义为：
+
+```
+   pub fn multiply_int_unsigned(
+
+      &self,
+
+      ctx: &mut EvalContext,
+
+      row: &[Datum],
+
+  ) -> Result<Option<i64>>
+```
+
+
+### Step 4: 实现函数逻辑
+
+这一块相对简单，直接对照 TiDB 的相关逻辑实现即可。这里，我们可以看到 TiDB 的 builtinArithmeticMultiplyIntUnsignedSig  的 具体实现如下：
+
+```
+func (s *builtinArithmeticMultiplyIntUnsignedSig) evalInt(row types.Row) (val int64, isNull bool, err error) {
+
+ a, isNull, err := s.args[0].EvalInt(s.ctx, row)
+
+ if isNull || err != nil {
+
+    return 0, isNull, errors.Trace(err)
+
+ }
+
+ unsignedA := uint64(a)
+
+ b, isNull, err := s.args[1].EvalInt(s.ctx, row)
+
+ if isNull || err != nil {
+
+    return 0, isNull, errors.Trace(err)
+
+ }
+
+ unsignedB := uint64(b)
+
+ result := unsignedA * unsignedB
+
+ if unsignedA != 0 && result/unsignedA != unsignedB {
+
+    return 0, true, types.ErrOverflow.GenByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%s * %s)", s.args[0].String(), s.args[1].String()))
+
+ }
+
+ return int64(result), false, nil
+
+}
+
+```
+
+参考以上代码，翻译到 TiKV 即可，如下：
+
+```
+pub fn multiply_int_unsigned(
```

Comment:
Please copy it from source https://github.com/pingcap/tikv/blob/master/src/coprocessor/dag/expr/builtin_arithmetic.rs#L169